include/deal.II/matrix_free/face_setup_internal.h

```diff
@@ -0,0 +1,944 @@
+        if (cell->has_periodic_neighbor(face_no))
+          info.exterior_face_no = cell->periodic_neighbor_face_no(face_no);
+        else
+          info.exterior_face_no = cell->neighbor_face_no(face_no);
+
+        info.subface_index = GeometryInfo<dim>::max_children_per_cell;
+        Assert(neighbor->level() <= cell->level(),
+               ExcInternalError());
+        if (cell->level() > neighbor->level())
+          {
+            if (cell->has_periodic_neighbor(face_no))
+              info.subface_index =
+                cell->periodic_neighbor_of_coarser_periodic_neighbor(face_no).second;
+            else
+              info.subface_index =
+                cell->neighbor_of_coarser_neighbor(face_no).second;
+          }
+
+        info.face_orientation = 0;
+        unsigned int  left_face_orientation =
+          !cell->face_orientation(face_no) + 2 * cell->face_flip(face_no) +
+          4 * cell->face_rotation(face_no);
+        unsigned int right_face_orientation =
+          !neighbor->face_orientation(info.exterior_face_no) +
+          2 * neighbor->face_flip(info.exterior_face_no) +
+          4 * neighbor->face_rotation(info.exterior_face_no);
+        if (left_face_orientation != 0)
+          {
+            info.face_orientation = 8 + left_face_orientation;
+            Assert(right_face_orientation == 0,
+                   ExcMessage("Face seems to be wrongly oriented from both sides"));
+          }
+        else
+          info.face_orientation = right_face_orientation;
+        return info;
+      }
+
+      bool use_active_cells;
+      std::vector<char> face_is_owned;
+      std::vector<bool> at_processor_boundary;
+      std::vector<unsigned int> cells_close_to_boundary;
+      std::vector<FaceToCellTopology<1> > inner_faces;
+      std::vector<FaceToCellTopology<1> > boundary_faces;
+      std::vector<FaceToCellTopology<1> > inner_ghost_faces;
+      std::vector<FaceToCellTopology<1> > refinement_edge_faces;
+    };
+
+
+    /**
+     * This simple comparison identifies faces of the same type, i.e., where
+     * all of the interior and exterior face number, subface index and
+     * orientation are the same. This is used to batch similar faces together
+     * for vectorization.
+     */
+    bool compare_faces_for_vectorization
+    (const FaceToCellTopology<1> &face1,
+     const FaceToCellTopology<1> &face2)
+    {
+      if (face1.interior_face_no != face2.interior_face_no)
+        return false;
+      if (face1.exterior_face_no != face2.exterior_face_no)
+        return false;
+      if (face1.subface_index != face2.subface_index)
+        return false;
+      if (face1.face_orientation != face2.face_orientation)
+        return false;
+      return true;
+    }
+
+
+
+    /**
+     * This comparator is used to create a sorting of FaceToCellTopology
+     * objects based on their identifiers. This is used to obtain a good data
+     * locality when processing the face integrals.
```

Comment:
Do I understand correctly that if `face1 == face2`, the comparator returns false.

Reply:
This is the usual `less_than` comparison used in sorting. If both the query `a < b ?` and `b < a ?` return `false`, the two values `a` and `b` are identical.